### DISCUSSION TITLE
PINTEREST: Wrong div loading

### DISCUSSION THREAD

**Discussion by cmeffley** [nss-evening-cohort-14/evening-client-side#178] [Help Me!] [answered]
Chris Meffley

No other issue tickets that I could see

Issue: When logging in or reloading page, the majority if the time, the pins div shows, not the boards. But, once in a blue moon, the boards div will load/show like is wanted. 

No error messages

https://toolset.com/forums/topic/on-webpage-refresh-browser-flicker/
https://stackoverflow.com/questions/6505419/stop-javascript-flicker-when-pages-content-is-manipulated
https://greensock.com/forums/topic/15482-flicker-on-first-load/
https://stackoverflow.com/questions/11966137/removing-the-page-flicker-when-loading-javascript-css

-I have double checked my domBuilder, making sure all divs have the proper syntax. 
-Checked the showBoards and showPins functions to make sure the correct divs are targeted and the others are empty. 
-The show cards by specific user is working correctly. 
-Commented out SCSS styles to see if that loading caused issues, it didn't seem to. 
-Also in SCSS tried visibility: initial for boards
-Tried window.onload for the board forEach loops functions

domBuilder: https://github.com/cmeffley/ASSIGNMENT-Firebase-Pinterest/blob/main/src/javascripts/components/domBuilder.js
showBoards: https://github.com/cmeffley/ASSIGNMENT-Firebase-Pinterest/blob/main/src/javascripts/components/boards.js
showPins: https://github.com/cmeffley/ASSIGNMENT-Firebase-Pinterest/blob/main/src/javascripts/components/pins.js
startApp: https://github.com/cmeffley/ASSIGNMENT-Firebase-Pinterest/blob/main/src/javascripts/views/startApp.js
HTML: https://github.com/cmeffley/ASSIGNMENT-Firebase-Pinterest/blob/main/src/index.html
GitHub Repo: https://github.com/cmeffley/ASSIGNMENT-Firebase-Pinterest

**Comment by drteresavasquez** [comment]
I'm not sure about this one. Look through your code and follow the events and the targets where you are putting your elements. Share the code blocks as you work your way through so that we don't have to ease though your code. Should be in the same file you're trying to load.

**Reply by cmeffley** [reply]
I went through the code and couldn't find any discrepancies. I'm still missing something. Maybe in the code blocks below, you might see something. I think what I did below is what you asked for. If not, I'll correct it.

**Comment by cmeffley** [comment]
domBuilder:
```Javascript
  document.querySelector('#app').innerHTML = `<div id="navigation"></div>
  <div id="main-container">
    <div id="title"></div>
    <div id="button"></div>
    <div id="form-container"></div>
    <div id="boards"></div>
    <div id="pins"></div>
  </div>`;
```
for showBoards:  
```document.querySelector('#title').innerHTML = '<h1>BOARDS</h1>';
  document.querySelector('#button').innerHTML = '<button class="btn btn-outline-light" id="add-board-btn">Add A Board</button>';
  document.querySelector('#form-container').innerHTML = '';
  document.querySelector('#boards').innerHTML = '';
  document.querySelector('#pins').innerHTML = '';

  array.forEach((item) => {
    document.querySelector('#boards').innerHTML += `<div class="card 
```

for showPins:
```document.querySelector('#title').innerHTML = '<h1>PINS</h1>';
  document.querySelector('#button').innerHTML = '<button class="btn btn-outline-light" id="add-pin-btn">Add A Pin</button>';
  document.querySelector('#form-container').innerHTML = '';
  document.querySelector('#boards').innerHTML = '';
  document.querySelector('#pins').innerHTML = '';

  array.forEach((element) => {
    document.querySelector('#pins').innerHTML += `<div class="card border-danger" id="pin-style">
```

for boardForm:
```const addBoardForm = () => {
  document.querySelector('#title').innerHTML = '';
  document.querySelector('#button').innerHTML = '';
  document.querySelector('#boards').innerHTML = '';
  document.querySelector('#pins').innerHTML = '';
  document.querySelector('#form-container').innerHTML = `<form id="new-board-form" class="mb-4">
  <div class="form-group">
```

for pinForm:
```  document.querySelector('#title').innerHTML = '';
  document.querySelector('#button').innerHTML = '';
  document.querySelector('#boards').innerHTML = '';
  document.querySelector('#pins').innerHTML = '';
  document.querySelector('#form-container').innerHTML = `<form id="new-pin-form" class="mb-4">
  <div class="form-group">
```

**Comment by cmeffley** [comment]
```if (e.target.id.includes('board-name-title')) {``` goes to ```  <a href="#"><h4 id="board-name-title--${item.firebaseKey}" ``` in showBoards

```  if (e.target.id.includes('delete-pin')) {``` goes to ``` <button class="btn btn-danger" id="delete-pin--${element.firebaseKey}">Delete Pin</button>``` in showPins

``` if (e.target.id.includes('delete-board')) {``` goes to ```   <button class="btn btn-danger" id="delete-board--${item.firebaseKey}">Delete Board</button>``` in showBoards

```  if (e.target.id.includes('add-board-btn')) {``` goes to ```  document.querySelector('#button').innerHTML = '<button class="btn btn-outline-light" id="add-board-btn">Add A Board</button>';``` in showBoards and emptyBoards

```if (e.target.id.includes('add-pin-btn')) {``` goes to ```  document.querySelector('#button').innerHTML = '<button class="btn btn-outline-light" id="add-pin-btn">Add A Pin</button>';``` in showPins and emptyPins

``` if (e.target.id.includes('submit-board')) {``` goes to ```  <button type="submit" id="submit-board" class="btn btn-outline-light">Submit Board</button>``` in addBoardForm

```   
board_title: document.querySelector('#new-board-name').value,
        image: document.querySelector('#boardPicture').value,
```
 goes to
```
    <input type="text" class="form-control" id="new-board-name" aria-describedby="boardName" placeholder="Enter Board Name" required><br/>
    <label for="basic-url" class="form-label">Add Picture</label>
    <input type="text" class="form-control" id="boardPicture" placeholder="Url" required><br/>
```
 in addBoardForm

```  if (e.target.id.includes('submit-pin')) {``` goes to ``` <button type="submit" id="submit-pin" class="btn btn-outline-light">Submit Pin</button>``` in addPinForm
```
        title: document.querySelector('#new-pin-name').value,
        image: document.querySelector('#pinPicture').value,
        board_id: document.querySelector('#select-board').value,
```
goes to
```
    <input type="text" class="form-control" id="new-pin-name" aria-describedby="pinName" placeholder="Enter Pin Name" required><br/>
    <label for="basic-url" class="form-label">Add Picture</label>
    <input type="text" class="form-control" id="pinPicture" placeholder="Url" required><br/>
    </div>
```
in addPinForm (board_id - see the last entry below, it shows where the 2 select-board go)

```if (e.target.id.includes('edit-pin-btn')) {``` goes to ```    <button class="btn btn-info" data-toggle="modal" data-target="#formModal" id="edit-pin-btn--${element.firebaseKey}">Change Boards</button>``` in showPins

``` if (e.target.id.includes('update-pin')) {``` goes to ```    <button type="submit" id="update-pin--${object.firebaseKey}" class="btn btn-primary">Update Pin</button>``` in editPinForm

``` board_id: document.querySelector('#select-board').value,``` goes to 
```
const selectBoard = (boardObject = {}) => {
  let domString = `<label for="board">Select a Board</label>
    <select class="form-control" id="select-board" required>
    <option value="">Select a Board</option>`;
```
in selectBoard

**Reply by drteresavasquez** [reply]
Dude. That's a lot of code. If your boards aren't showing up it means that you are either clearing out your div with something else or you're not targeting your div properly. Comment out all the other stuff except boards ON A NEW BRANCH and then see if they load. If they don't you should be able to easily find your problem. If they do, look at the components that load after it one by one and see which one is causing the problem.

**Reply by cmeffley** [reply]
Earlier I did comment out just getAllBoards in startApp and the pins AND boards showed. Then I commented out just getAllPins and the boards AND pins showed. I could not figure out why. It's something I will revisit. My brain hurts, I'll look at it with fresh eyes tomorrow.

**Comment by drteresavasquez** [comment]
Also what does your postman GET response look like for your user Id? Do you have boards in Firebase with that on it?

**Reply by cmeffley** [reply]
Hopefully this will clear things up:
https://www.loom.com/share/c87e1e0710e343b3811f664c39d13446

**Comment by drteresavasquez** [comment]
On line 19 in start app you call get all boards and then after that you call get all pins. In the show pins function you clear out the board inner html 🤷🏽‍♀️ that's why you can't see boards.

**Reply by cmeffley** [reply]
When I comment out the board innerHTML in the showPins, it shows all the pins, but it leaves the boards on the screen as well.

**Reply by drteresavasquez** [reply]
Because you are calling both in start app! And they are targeting separate divs. So I would say the answer is found!! 🎉🎉

**Reply by drteresavasquez** [reply]
Delete/comment out this code and boards will load...pins will not. https://github.com/cmeffley/ASSIGNMENT-Firebase-Pinterest/blob/main/src/javascripts/views/startApp.js#L26-L33

**Comment by drteresavasquez** [comment]
Follow up on this. @cmeffley if you've moved on or found a solution, please update the ticket with a comment. Thanks.

**Reply by cmeffley** [reply]
I just jumped on to try and hopefully work on a solution

**Comment by cmeffley** [comment, marked as answer]
Ok, I think the light bulb has finally gone off in my head...... I combined the boards and pins divs into one div (see-cards), to make sure there is no overriding of divs because of the cascading affect. In startApp, since I was calling the getAllBoards function first, the DOM quickly showed the boards. But then I called the getAllPins function next, so that overrode the getAllBoards function so the pins would show, again, because of cascading affect. So it was unnecessary to call the getAllPins function in the start app.
